services/gateway_manager/src/main/java/com/futurewei/alcor/gatewaymanager/service/GatewayService.java

```diff
@@ -0,0 +1,188 @@
+
+    @Autowired
+    private GatewayManagerRestClinet restClinet;
+
+    private static final String GATEWAY_DESCRIPTION = "internal gateway";
+    private static final String GATEWAY_SUFFIX_NAME = " name";
+    private static final String ATTACHMENT_NAME_PREFIX = "attachment - ";
+
+
+    public GatewayService() {
+    }
+
+    public GatewayInfo createGatewayInfo(String projectId, VpcInfo vpcInfo) throws Exception {
+
+        //TODO check whether the project type is zeta by projectId
+
+        // create GatewayInfo entity in the DB
+        GatewayInfo gatewayInfo = new GatewayInfo();
+        GatewayEntity gatewayEntity = new GatewayEntity();
+
+        // construct the GatewayEntity and Attachment
+        createGatewayAndAttachment(projectId, vpcInfo, gatewayInfo, gatewayEntity);
+
+        AsyncExecutor executor = new AsyncExecutor();
+        executor.runAsync(restClinet::createDPMCacheGateway, projectId, gatewayInfo);
+        executor.runAsync(restClinet::createVPCInZetaGateway, new VpcInfoSub(vpcInfo.getVpcId(), vpcInfo.getVpcVni()));
+        executor.runAsync((Supplier<Void>) () -> {
+            // wait result of all async
+            List<Object> result = executor.joinAllAsync();
+            // whether to rollback by checking result's size is 2
+            try {
+                updateDPMCacheGateway(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("update GatewayEntity status to READY failed, detail message: {}", e.getMessage());
+                //TODO how to handle this exception? rollback all?
+            }
+
+            try {
+                rollback(result, gatewayInfo, gatewayEntity, projectId);
+            } catch (Exception e) {
+                log.info("rollback failed, detail message: {}", e.getMessage());
+                //TODO if rollback failed,how we should handle it?
+            }
+            return null;
+        });
+        return gatewayInfo;
+    }
+
+    public void updateGatewayInfoForZeta(String projectId, GatewayInfo newGatewayInfo) throws Exception {
+        //TODO check whether the project type is zeta by projectId
+
+        GatewayEntity gatewayEntity = null;
+        Map<String, GWAttachment> attachmentsMap = gwAttachmentRepository.findAllItems();
+        for (GatewayEntity newGatewayEntity : newGatewayInfo.getGatewayEntities()) {
+            for (GWAttachment attachment : attachmentsMap.values()) {
+                if (attachment.getResourceId().equals(newGatewayInfo.getResourceId())) {
+                    gatewayEntity = gatewayRepository.findItem(attachment.getGatewayId());
+                    if (gatewayEntity == null) {
+                        throw new Exception(ExceptionMsgConfig.GATEWAY_ENTITY_NOT_FOUND.getMsg());
+                    }
+                    if (gatewayEntity.getType().equals(newGatewayEntity.getType())) {
+                        gatewayEntity.setStatus(newGatewayEntity.getStatus());
```

Comment:
need gatewayRepository.addItem(gatewayEntity) here ?

Reply:
Have updates in the code below